### ISSUE TITLE
Pod still running after being force deleted with 0 grace period, network policices stop applying

### ISSUE THREAD

**Issue by mierak** [kubernetes/kubernetes#110507]
### What happened?

Force deleting a pod with grace-period=0 does not immedietally stop it. Instead the pod disappears, is not listed anywhere. Only way to check if the pod is doing something is to access log files directly on the node fs.
Additionally, all network policies stop working when the pod gets in this half-dead state.

### What did you expect to happen?

The pod gets killed immedietally

### How can we reproduce it (as minimally and precisely as possible)?

* Create a deployment, 
* Force delete the created pod (kubectl delete pods <POD_NAME> --grace-period=0 --force)
* Check log file of the pod on node FS(or some other way, we used promtails mount to logs)
* It is still there and being actively updated until the pod gets killed by terminationGracePeriodSeconds

```apiVersion: apps/v1
kind: Deployment
metadata:
  name: shutdowntest-deployment
  labels:
    app: shutdowntest
spec:
  replicas: 1
  selector:
    matchLabels:
      app: shutdowntest
  template:
    metadata:
      labels:
        app: shutdowntest
    spec:
      containers:
        - name: shutdowntest
          image: redhat/ubi8-minimal
          command: ["/bin/sh"]
          args:
            [
              "-c",
              "for i in $(seq 0 300); do echo still alive $i; sleep 1; done",
            ]
      terminationGracePeriodSeconds: 360```

### Anything else we need to know?

Once the pod gets into this half terminated state, it is not visible anymore and network policies stop working. For example, if the half-dead pod is calling another, which has specifically enabled ingress from our subject, it cannot be reached anymore.

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.5", GitCommit:"5c99e2ac2ff9a3c549d9ca665e7bc05a3e18f07e", GitTreeState:"clean", BuildDate:"2021-12-16T08:38:33Z", GoVersion:"go1.16.12", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.6", GitCommit:"07959215dd83b4ae6317b33c824f845abd578642", GitTreeState:"clean", BuildDate:"2022-03-30T18:28:25Z", GoVersion:"go1.16.12", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by liggitt** [comment]
cc @smarterclayton 

Don't delete with a zero second grace period if you want an orderly tear down and for things like network policies to continue working during tear down. grace period of 0 and --force are saying "remove the record of this pod from the API, even before the kubelet has completed teardown"

**Comment by vbystricky21** [comment]
@liggitt you are right. Not working network policy is not our primary concern. It is the fact that the pod we thought should be terminated pretty much instantly (with grace period of 0) is running for the whole duration of terminationGracePeriodSeconds.

**Comment by brianpursley** [comment]
When using kubectl, `--force` and `--grace-period=0` are not the same thing.  If you set `--grace-period=0`, then kubectl changes it to a grace period of 1, as shown below:

https://github.com/kubernetes/kubernetes/blob/fdb2d544751adc9fd2f6fa5075e9a16df7d352df/staging/src/k8s.io/kubectl/pkg/cmd/delete/delete.go#L180-L193

So the options affecting grace period are:
* `--force` results grace period of 0.  UNLESS... you also specify a grace period, then it is ignored and the specified grace period is used (I think).
* `--grace-period=0` unexpectedly results in grace period of 1.
* `--grace-period=N` where N > 0, does that you expect, resulting in grace period of N.
* `--now` is an alias for grace period of 1.

See also https://github.com/kubernetes/kubernetes/issues/108741
